### PR TITLE
Add basic fuzzer

### PR DESCRIPTION
**Pull request by devatwork**
Basic XPath expression fuzzer inspired by @gamozolabs

This script implements a simple fuzzer which uses a `corpus` of valid XPath expressions which it mutates using `mutateCharactersInPlace` and `mutateString` and then evaluates to test the parser and the expression engine for bugs.

The fuzzer deduplicates errors thrown by their unique stack trace and filters out expected XPath errors which are properly generated by the engine to reduce noise.

The fuzzer reports it's progress, including the number of unique crashes and the number of fuzz cases per second (`fcps`) every 1000 cases.

# Running the fuzzer

```
npm run fuzzer
```

# Ideas/Improvements

* Multi CPU core support
* Component library + mutation strategy
* Clean the corpus
* Corpus feedback
* Run fuzzer on compiled XPath engine
* Disable parse cache

### CODE REVIEW

**Comment by devatwork** [comment]
Example crashes found by the fuzzer:

```
!!! Found error !!
Input: ancestor::*!name() = 'app' or name() = 'sec' or name() = 'term-sec'![1]
TypeError: Cannot read property 'createSequenceIterator' of null
    at SimpleMapOperator.evaluate (C:\fontoxml\platform\fontoxpath\src\expressions\operators\SimpleMapOperator.ts:33:47)
    at SimpleMapOperator.evaluateWithoutFocus (C:\fontoxml\platform\fontoxpath\src\expressions\Expression.ts:109:10)
    at SimpleMapOperator.evaluateMaybeStatically (C:\fontoxml\platform\fontoxpath\src\expressions\Expression.ts:75:16)
    at Compare.evaluate (C:\fontoxml\platform\fontoxpath\src\expressions\operators\compares\Compare.ts:54:49)
    at Compare.evaluateMaybeStatically (C:\fontoxml\platform\fontoxpath\src\expressions\Expression.ts:77:15)
    at Object.next (C:\fontoxml\platform\fontoxpath\src\expressions\operators\boolean\OrOperator.ts:79:39)
    at Object.next (C:\fontoxml\platform\fontoxpath\src\expressions\dataTypes\Sequences\IteratorBackedSequence.ts:30:33)
    at IteratorBackedSequence.getAllValues (C:\fontoxml\platform\fontoxpath\src\expressions\dataTypes\Sequences\IteratorBackedSequence.ts:103:22)
    at Object.convertXDMReturnValue [as default] (C:\fontoxml\platform\fontoxpath\src\parsing\convertXDMReturnValue.ts:233:33)
    at Object.evaluateXPath (C:\fontoxml\platform\fontoxpath\src\evaluateXPath.ts:188:41)
```

```
!!! Found error !!
Input: 5!s81
TypeError: Cannot read property 'nodeType' of undefined
    at DomFacade.getNodeType (C:\fontoxml\platform\fontoxpath\src\domFacade\DomFacade.ts:252:23)
    at ChildAxis.evaluate (C:\fontoxml\platform\fontoxpath\src\expressions\axes\ChildAxis.ts:30:30)
    at ChildAxis.evaluateMaybeStatically (C:\fontoxml\platform\fontoxpath\src\expressions\Expression.ts:77:15)
    at Object.next (C:\fontoxml\platform\fontoxpath\src\expressions\operators\SimpleMapOperator.ts:50:49)
    at Object.next (C:\fontoxml\platform\fontoxpath\src\expressions\dataTypes\Sequences\IteratorBackedSequence.ts:30:33)
    at IteratorBackedSequence.getAllValues (C:\fontoxml\platform\fontoxpath\src\expressions\dataTypes\Sequences\IteratorBackedSequence.ts:103:22)
    at Object.convertXDMReturnValue [as default] (C:\fontoxml\platform\fontoxpath\src\parsing\convertXDMReturnValue.ts:233:33)
    at Object.evaluateXPath (C:\fontoxml\platform\fontoxpath\src\evaluateXPath.ts:188:41)
    at worker (C:\fontoxml\platform\fontoxpath\fuzzers\fuzzer.ts:355:4)
    at Object.<anonymous> (C:\fontoxml\platform\fontoxpath\fuzzers\fuzzer.ts:386:1)
```

```
!!! Found error !!
Input: let $vontext := ., $listAnce:tor := $context/ancestor-or-self::*[fonto:list(.)][1], $closedAncestor := $context/ancestor-or-self::*[fonto:closed(.)][1] return if ($listAncestor and not($closOdAncestor)) then true() else $listAncestor >> $closedAncestor
Error: Not implemented: let expressions with namespace usage.
    at new LetExpression (C:\fontoxml\platform\fontoxpath\src\expressions\LetExpression.ts:37:10)
    at letClause (C:\fontoxml\platform\fontoxpath\src\parsing\compileAstToExpression.ts:484:16)
    at C:\fontoxml\platform\fontoxpath\src\parsing\compileAstToExpression.ts:573:13
    at Array.reduceRight (<anonymous>)
    at flworExpression (C:\fontoxml\platform\fontoxpath\src\parsing\compileAstToExpression.ts:563:17)
    at compile (C:\fontoxml\platform\fontoxpath\src\parsing\compileAstToExpression.ts:172:11)
    at Object.default_1 [as default] (C:\fontoxml\platform\fontoxpath\src\parsing\compileAstToExpression.ts:1418:9)
    at Object.staticallyCompileXPath [as default] (C:\fontoxml\platform\fontoxpath\src\parsing\staticallyCompileXPath.ts:78:38)
    at Object.buildEvaluationContext [as default] (C:\fontoxml\platform\fontoxpath\src\evaluationUtils\buildEvaluationContext.ts:132:59)
    at Object.evaluateXPath (C:\fontoxml\platform\fontoxpath\src\evaluateXPath.ts:145:41)
```
Probably intentional but maybe should have an `XP` something prefix?

**Comment by DrRataplan** [comment]
Cool start! Nice to see it already found some random crashes. I'll take a minute or two to fix the CodeFactor errors.

I also would like to put this into the CI. If it runs automatically for new commits for a few seconds, the chance is high it will find more things in the future. I'll keep it excluded for code coverage due to its random nature.

**Comment by devatwork** [comment]
Thanks, I spend most of the type to get the thing to build in the first place, I don't know if the setup is any good so feel free to make changes or tell me what to do.

The issues if found took about ~30 minutes to find the current issues, so I'm not sure if running it in the CI makes too much sense (altough it will act as a sort of smoke test). Of course we can add the cases it found so far to the corpus and there is plenty of room for optimization. But the fuzzer is basically brute forcing.

Do you want me to add a section to the README/contributing guide as well?

**Comment by DrRataplan** [comment]
A section to the CONTRIBUTING would be ideal.

When I ran the fuzzer locally, I had an error within minutes, but I think it makes more sense to first fix these errors before integrating, to prevent broken build almost immediately.

I fixed one of those errors, just a wrong path being taken resulting in a type error instead of an `XPDY0002`. Making a separate PR of that. Once this is in, I'll run the fuzzer for a significant time and I'll just log the errors as issues.

I also made the fuzzer work on the built version. I think it makes more sense that way because we _have_ seen the typical issues a fuzzer would find in built versions only. thanks to the Closure compiler renaming variables we forgot the mark as unrenamable.

Thanks again!

**Comment by devatwork** [comment]
> A section to the CONTRIBUTING would be ideal.
Ok cool, I'll add someting.

> When I ran the fuzzer locally, I had an error within minutes, but I think it makes more sense to first fix these errors before integrating, to prevent broken build almost immediately.
It is random, so yes might happen on the first try, might happen on the millionth time, no guarantees.

> I fixed one of those errors, just a wrong path being taken resulting in a type error instead of an `XPDY0002`. Making a separate PR of that. Once this is in, I'll run the fuzzer for a significant time and I'll just log the errors as issues.
Ok cool, glad to see that it found actual bugs.

> I also made the fuzzer work on the built version. I think it makes more sense that way because we _have_ seen the typical issues a fuzzer would find in built versions only. thanks to the Closure compiler renaming variables we forgot the mark as unrenamable.
Sure make sense to me, that should help perf. The stack trace the fuzzer reports may not be useful but it also outputs the expression which causes the crash, so it should be easy to repro in debug for further analysis which is how it should be. I'll make sure to mention that in the contrib guide.

> Thanks again!
You're welcome, it was a fun project for a few hrs. It's my first fuzzer and it wat surprisingly easy to write. Most time went into figuring out the wiring for the build, not the code itself. I'll include some more links to video material that I found useful for others if they're interested.

**Comment by DrRataplan** [comment]
For that cache disabling, we have a special `disableCache` option: https://github.com/FontoXML/fontoxpath/blob/master/src/evaluateXPath.ts#L157. It's internal, but it is not renamed to something random like the rest of the internal variables / arguments.

**Comment by DrRataplan** [comment]
Also, I just merged #335, which includes a better tslint.json that you can _also_ use in this `fuzzer` branch to get TSLint / CodeFactor of your back for those imports.

File: fuzzers/fuzzer.ts [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+
+			// Print the error
+			process.stdout.write(
+				`\n\n!!! Found error !!\nInput: ${expression}\n${error.stack}\n\n`
+			);
+		}
+	}
+}
+
+// TODO: Figure out how to scale this to multiple cpu cores to fuzz faster
```
Comment (DrRataplan) [inline fuzzers/fuzzer.ts:384]: `npm run fuzzer & npm run fuzzer & npm run fuzzer &`? :laughing:

**Review by DrRataplan — APPROVED** [review]

File: fuzzers/fuzzer.ts (hunk 1 above)
Comment (devatwork) [reply fuzzers/fuzzer.ts:384]: I have 4 cores+HT so your command would not utilize my CPU to it's fullest potential 🙃. It shouldn't be too hard (I hope) to get this done properly in nodejs.

**Comment by devatwork** [comment]
Thanks for you help @DrRataplan. I'll hope I can make the remaining changes soonish:

- [ ] Add instructions for the fuzzer to CONTRIBUTING.
- [ ] Update tslint.json to #335.
- [ ] Use `disableCache` option to disable, unsurprisingly, the cache.
- [ ] Include references to interesting reading material.

**Comment by DrRataplan** [comment]
## BundleMon

  
  
  
  
  
  
  
<details>
<summary>Unchanged files (2)</summary>

Status | Path | Size | Limits
:------------: | ------------ | :------------: | :------------:
:white_check_mark: | fontoxpath.esm.js<br/> | 83.79KB | -
:white_check_mark: | fontoxpath.js<br/> | 83.69KB | -
</details>



No change in files bundle size

Final result: :white_check_mark:

---
<p align="center"><a href="https://bundlemon.now.sh/projects/609024effb87a20009a6b096/reports?branch=fuzzer&resolution=all" target="_blank" rel="noreferrer noopener">Current branch size history</a> | <a href="https://bundlemon.now.sh/projects/609024effb87a20009a6b096/reports?branch=master&resolution=days" target="_blank" rel="noreferrer noopener">Target branch size history</a></p>

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -199,6 +199,18 @@ storing these in a file you can run the same set even when switching between dif
 the generated file present you can run the tests using `npm run qt3performance`, this will run a
 benchmark for each qt3 test using [benchmarkjs](https://benchmarkjs.com/).
 
+### Running the fuzzer
+
+FontoXPath comes with a fuzzer which to find bugs automatically. The fuzzer has a corpus of ~200 XPath expressions which are randomly mutated.
```
Comment (DrRataplan) [inline CONTRIBUTING.md:204]: FontoXPath comes with a fuzzer *with* which to find bugs automatically.

or

FontoXPath comes with a fuzzer that can find bugs automatically.

**Review by DrRataplan — APPROVED** [review]

**Comment by devatwork** [comment]
@DrRataplan I implemented fuzzing for the 3 different language dialects supported. There is plenty more that we could do but I guess it is good enough for now. Agree?

**Pull request merged**